OpenCL_C.txt

```diff
@@ -119,7 +119,8 @@ by the presence of special predefined macros.
 === Features
 
 IMPORTANT: Feature test macros <<unified-spec, require>> support for OpenCL C
-3.0 or newer.
+3.0 or newer. This means that applications which target previous OpenCL C versions
```

Comment:
Generally, double underscore prefixed identifiers are reserved in C, and OpenCL doesn't restrict this. This means there is no harm for implementations to define such macros in earlier OpenCL C versions i.e. there shouldn't be name clusing with the application code.

So I agree with your approach and I think it will make things easier in clang if we add them unconditionally. 

Regarding the phrasing in the spec perhaps we can refer to C99 section 7.1.3  "Reserved identifiers" paragraph 1

` All identifiers that begin with an underscore and either an uppercase letter or anotherunderscore are always reserved for any use.`

stating that implementations might provide (but don't have to!) definitions of these macros in earlier OpenCL C versions?

Reply:
> All identifiers that begin with an underscore and either an uppercase letter or anotherunderscore are always reserved for any use

Well, looking from this point of view we are definitely fine to define feature test macros whenever we want and especially for pre-OpenCL C 3.0. I'm just wondering if special adjustment needed for OpenCL C spec because of this:

> _IMPORTANT_: Feature test macros support _for OpenCL C 3.0 or newer_.

Reply:
>    IMPORTANT: Feature test macros require support for OpenCL C 3.0 or newer.

Ok, this leaves the behavior of earlier versions unspecified.

The way I see OpenCL C spec is that it intends to document the difference to C99 used as a base. That implies whenever OpenCL doesn't specify certain behavior C99 rules apply.

If we consider the above then we probably don't need to add anything strictly speaking. However, I am very much in favor of self-contained specs because they reduce the pain of reply on separate documents and they also tend to be more consistent. So I find it helpful where possible to provide extra clarifications if they can be made concisely. The issues you have highlighted seems to be a good candidate. However, I think it would help to reference C99 to make it clear we are not reinventing the wheel and also to point out where more information is to be found.

One thing we should consider though that the macros with the above names could `theoretically` be already used by the implementations for different purposes from those documented in the spec. It makes the matter a bit more complicated but however this is highly unlikely.

Reply:
>"Feature test macros are only available in OpenCL C 3.0 or newer".

We discussed it with @bashbaug and its purpose was as exactly as I wrought down in this patch.

>The way I see OpenCL C spec in general is that it intends to document the difference to C99 used as a base. That implies whenever OpenCL doesn't specify certain behavior C99 rules apply.

Yes, but here spec specifies this behavior.

> However, I am very much in favor of self-contained specs because they reduce the pain of relying on separate documents and they also tend to be more consistent

Yeah, I would be glad to write some more clarifications, something like: _compilers for earlier OpenCL C versions are allowed to define feature test macros, but they are not required doing so_, if spec maintainers will find it valuable. I personally don't think it will hurt in any cases except that there are some other causes of putting this IMPORTANT note in the spec.